### PR TITLE
Fix URI comparisons for different casing

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2208409

Issue was caused by https://github.com/dotnet/roslyn/pull/74544

Prior to my change, absolute URIs would ignore case when comparing UNC URIs (default impl of `Uri.GetHashCode`.  However once I manually started comparing AbsolutePaths, case was never ignored (due to the hashcode being different for upper vs lower case strings).

This showed up as an issue in C# DevKit because files in the workspace get added with an upper case drive letter (by devkit project system).  But VSCode sends LSP requests always with a lower case drive letter.  We'd end up always ignoring the diagnostics requests because we thought the file was not tracked by LSP.

This did not show up as an issue in the C# extension because we populate the workspace there with lower case drive letters (what vscode gives us).

The change here is to use an ignoring case hashcode for absolute path.  This will potentially cause collisions if a path exists with both cases, but that is generally rare.  We will fallback to the URI implementation of comparison in `Equals`, which appropriately handles case depending on the kind of URI.

### CODE REVIEW

File: src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs [hunk 1]
```diff
@@ -76,10 +76,19 @@ public int GetHashCode(Uri obj)
...
+                // However, depending on the kind of URI, case sensitivity in AbsoluteUri should be ignored.
+                // Uri.GetHashCode normally handles this internally, but the parameters it uses to determine which comparison to use are not exposed.
+                //
+                // Instead, we will always create the hash code ignoring case, and will rely on the Equals implementation
+                // to handle collisions (between two Uris with different casing).  This should be very rare in practice.
+                return StringComparer.OrdinalIgnoreCase.GetHashCode(obj.AbsoluteUri);
             }
             else
             {
+                obj.GetComponents(UriComponents.AbsoluteUri, UriFormat.UriEscaped);
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs:91]: what does this logic do?  it seems to be ignroed.

Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs:91]: > This should be very rare in practice.

consider giving an example of where it does arise.

File: src/LanguageServer/ProtocolUnitTests/UriTests.cs [hunk 2]
```diff
@@ -210,6 +210,91 @@ public async Task TestFindsExistingDocumentWhenUriHasDifferentEncodingAsync(bool
         Assert.Same(originalText, encodedText);
     }
 
+    [Theory, CombinatorialData]
```
Comment (CyrusNajmabadi) [inline src/LanguageServer/ProtocolUnitTests/UriTests.cs:213]: no workitem?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs (hunk 1 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs:91]: oops, left that there accidentally.

**Review by JoeRobich — APPROVED** [review]

**Pull request merged**